src/python/infer_token_vectors_cohere.py

```diff
@@ -1,66 +1,130 @@
+  parser = argparse.ArgumentParser(prog='Fetch Wikipedia Cohere Embeddings',
+                                     description='Generate document and query vectors for the vector search task '
+                                                 'from HuggingFace Cohere/wikipedia-22-12-en-embeddings')
+  parser.add_argument('-n', '--name', default='cohere-wikipedia',
+                      help='Dataset name, used as a filename prefix for generated files.')
+  parser.add_argument('-d', '--numDocs', default='1_000_000', help='Number of documents')
+  parser.add_argument('-q', '--numQueries', default='10_000', help='Number of queries')
+  args = parser.parse_args()
+  print('Fetching Cohere embeddings with the following args: %s' % args)
+
+  doc_file = f"{localconstants.BASE_DIR}/data/{args.name}-docs-{DIMENSIONS}d.vec"
+  query_file = f"{localconstants.BASE_DIR}/data/{args.name}-queries-{DIMENSIONS}d.vec"
+  meta_file = f"{localconstants.BASE_DIR}/data/{args.name}-metadata.csv"
+  num_docs = int(args.numDocs)
+  num_queries = int(args.numQueries)
+
+  for name in (doc_file, query_file, meta_file):
+    print(f'checking if file:{name} exists...')
+    if os.path.exists(name):
+        raise RuntimeError(f'please remove {name} first')
+
+  ds = datasets.load_dataset(DATASET_PATH, split="train")
+  print(f'features: {ds.features}')
+  print(f"total number of rows: {len(ds)}")
+  embedding_dims = len(ds[0]['emb'])
+  print(f"embeddings dims: {embedding_dims}")
+  assert embedding_dims == DIMENSIONS , f'Dataset embedding dimensions: {embedding_dims} do not match configured dimensions: {DIMENSIONS}'
+
+  with open(meta_file, "at") as meta:
+    meta.write(f'wiki_id,para_id\n')
+
+  # do this in windows, else the RAM usage is crazy (OOME even with 256
+  # GB RAM since I think this step makes 2X copy of the dataset?)
+  doc_upto = 0
+  window_num_docs = 1000000
+  # window_num_docs = 100000
+
+  # Fetch Document Embeddings
+  while doc_upto < num_docs:
+    next_doc_upto = min(doc_upto + window_num_docs, num_docs)
+    ds_embs = ds[doc_upto:next_doc_upto]['emb']
+    ds_wiki_id = ds[doc_upto:next_doc_upto]['wiki_id']
+    ds_para_id = ds[doc_upto:next_doc_upto]['paragraph_id']
+    batch_size = next_doc_upto - doc_upto
+    print(f'batch size = {batch_size}')
+    #print(f'\nds_embs: {ds_embs}')
+    embs = np.array(ds_embs, dtype=np.float32)
+    print(f'embs: {embs.dtype} {embs.size} {embs.itemsize} {embs.shape}')
+    print(f'wiki_id: {len(ds_wiki_id)}')
+    print(f'para_id: {len(ds_para_id)}')
+
+    print(f"saving docs[{doc_upto}:{next_doc_upto}] of shape: {embs.shape} to file")
+    with open(doc_file, "ab") as out_f:
+        embs.tofile(out_f)
+
+    print(f'writing associated metadata')
+    with open(meta_file, "at") as meta:
+      for i in range(batch_size):
+        meta.write(f'{ds_wiki_id[i]},{ds_para_id[i]}\n')
+
+    doc_upto = next_doc_upto
+
+  # Fetch Query Embeddings
+  ds_embs_queries = ds[num_docs : num_docs + num_queries]['emb']
```

Comment:
I'm a bit iffy about this. 
The dataset tends to have ~100 sub-documents (paragraphs) per parent (wiki_id). By sequentially picking the next N embeddings as queries, we are only getting ~N/100 diverse query topics. A lot of queries will end up being different paragraphs (sub-docs) in the same document and pick the same parent. I wonder what this does to recall metrics...

Reply:
Maybe we could randomly or deterministically sample every Mth embedding of what remains after we sucked out `num_docs` for indexing?  Can come as a later improvement...